src/codegen/cfg.rs

```diff
@@ -185,6 +192,9 @@ pub enum Instr {
     },
     /// Return a code at the end of a function
     ReturnCode { code: ReturnCode },
+    /// Report the runtime error before reaching aborting execution (AssertFailure).
+    /// If Instr::Print is used, the error string and location must be turned to an Expression first.
+    ReportError { string: String, loc: Loc },
```

Comment:
Now we have `Instr::Print` and `Instr::ReportError`, as well as `fn print()` and `fn debug_print()` in emit. But they do the same thing, it is unlikely we ever going to have two possible locations where something could be printed into (at least for the substrate target(. Looking at the implementation, it is not clear to me why `Instr::ReportError` is needed, the two isntructions do almost exactly the same. Also I feel like this will make it awkward for formatting the debug buffer. Can you share your thoughts behind having a dedicated instruction for this?

Reply:
Other than that, this doesn't look to bad @salaheldinsoliman! I really like the many test cases you added :medal_sports:

Reply:
My rationale is more readability: If Instr::Print is used, either an extra parameter should be added (Option<String>) maybe or convert its parameter to an enum having an expression and string variations. Also ``loc`` has to be passed along to be concatenated to the error. I think maintenance in this case would be harder... 
I have thought about this before starting, and read this nice [thread](https://softwareengineering.stackexchange.com/questions/194504/should-my-code-be-dry-or-readable-if-it-cant-be-both) and reached my final verdict of making an extra instruction

Reply:
I see, however I don't think an extra instruction with code added to `emit` makes the code more readable here. If you move the logic of `fn print_runtime_error()` into codegen, you don't have to deal with passing `String`s and `Loc`s around at all.

Reply:
I considered making a codegen expression from the error string and loc, but this would make the CFG look a bit complicated if someone wants to debug his code using ``--emit cfg``

Reply:
Other than that CFG matter, I think removing Instr::ReportError all together is much better.

Reply:
How is this going to add substantially to the CFG? And if this should ever be an issue, just don't use `--log-runtime-errors` to debug the CFG, problem solved.

The `print_runtime_error` can exist in emit without having a  `ReportError` instruction. It could also be made into a callable function, then you should be able to refactor this into codegen and just call it in emit. 

I'm still not convinced introducing an additional `ReportError` instruction is justified. @seanyoung WDYT?

Reply:
I had another look today and you are correct. repetition here can be avoided easily if a string is converted to an expression. a programmer debugging the CFG will now that this is inserted by the compiler because of (``Loc::Codegen``). Thanks!
I will push a commit removing ReportError all together and move ``print_runtime_error`` to codegen.